### PR TITLE
Feature/conveyor/conveyor implementation

### PR DESCRIPTION
**Pull request by cblpope3**
Модуль Conveyor (по пункту 3.1).

### CODE REVIEW

**Comment by cblpope3** [comment]
Пока неправильно работает расчёт ПСК и нет юнит-тестов.

File: Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        credit.setTerm(creditTerm);
+        credit.setRate(creditRate);
+        credit.setIsInsuranceEnabled(isInsuranceEnabled);
+        credit.setIsSalaryClient(isSalaryClient);
+
+        // размер ежемесячного платежа(monthlyPayment),
+        BigDecimal monthlyPayment = this.calculateMonthlyPayment(creditAmount, creditRate, creditTerm);
+        // график ежемесячных платежей (List<PaymentScheduleElement>)
+        List<ModelsPaymentScheduleElementDTO> paymentSchedule = this.calculateMonthlyPaymentsSchedule(
+                creditAmount, creditTerm, creditRate.divide(BigDecimal.valueOf(1200), MathContext.DECIMAL64), monthlyPayment);
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java:108]: Числа и любые подобные значения (строки) в коде лучше не писать в таком виде, всегда лучше указывать их в константах сверху, потому что так можно задать некий контекст, что это за поле, через ее имя

Comment (alxdv97) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java:108]: Например в данном случае, тот, кто будет потом читать этот код должен будет потратить несколько минут, на осознание того факта, что 1200 - это 12 месяцев и 100 процентов в одном поле))). То есть 1200 нужно во-первых разделить на 2 честных поля (месяцы и проценты), во-вторых, объявить как константы

File: Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        // PSK = 12 * 100 * ((s/s0) - 1) / n
+        // where
+        // s - total payments sum
+        // s0 - credit amount
+        // n - number of months of credit term
+
+        //fixme calculation result is not matching with online-calculator result
+        BigDecimal totalPaymentsSum = paymentSchedule.stream()
+                .map(ModelsPaymentScheduleElementDTO::getTotalPayment)
+                .reduce(BigDecimal.valueOf(0),
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java:138]: Для числе 0, 1, 10 есть отличный статический метод у BigDecimal - BigDecimal.ZERO, BigDecimal.ONE, BigDecimal.TEN

File: Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+                        BigDecimal::add);
+
+        BigDecimal overpayCoefficient = totalPaymentsSum.divide(creditAmount, MathContext.DECIMAL64);
+        BigDecimal partialResult = overpayCoefficient.subtract(BigDecimal.valueOf(1))
+                .divide(BigDecimal.valueOf(creditTerm), MathContext.DECIMAL64);
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java:143]: если я не ошибаюсь, можно не указывать MathContext.DECIMAL64

File: Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+        // Также принимаем partialResult = monthlyCreditRate/(overpayCoefficient - 1)
+        BigDecimal partialResult = monthlyCreditRate
+                .divide(overpayCoefficient.subtract(BigDecimal.valueOf(1), MathContext.DECIMAL64),
+                        MathContext.DECIMAL64);
+
+        // Тогда monthlyPayment = creditAmount*(monthlyCreditRate+partialResult)
+        return creditAmount.multiply(
+                        monthlyCreditRate.add(partialResult, MathContext.DECIMAL64), MathContext.DECIMAL64)
+                .setScale(2, RoundingMode.HALF_UP);
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java:248]: Со скейлом надо внимательно посмотреть по всему коду, потому что скейл 2 важен только для финального числа, которые пойдет в ответ, а внутри системы, для некоторых коэффициентов иногда нужно скейл чуть ли не 8-10 делать.
Может из-за этого что-то не сходится в расчетах ПСК - там могут быть очень маленькие промежуточные числа

File: Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+import java.time.Period;
+import java.util.List;
+
+/**
+ * Service that perform pre-scoring job.
+ */
+@Service
+public class PreScoringService {
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java:23]: См. комменты выше про ломбок

File: Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+     * This method calculating four credit offers to customer.
+     *
+     * @param loanRequest requested credit.
+     * @return {@link List} of four credit offers.
+     * @throws LoanRequestException if request not passed pre-verification process.
+     */
+    public List<ModelsLoanOfferDTO> getCreditOfferList(ModelsLoanApplicationRequestDTO loanRequest) throws LoanRequestException {
+
+        //checking request. exception will be thrown if issues happened.
+        this.validateLoanApplicationRequestDTO(loanRequest);
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java:45]: Обращение через this в данном контексте вроде не нужно. Оно нужно только тогда, когда у тебя в текущем скоупе есть еще один метод/поле с таким же названием

File: Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    /**
+     * Method checks that loan request has valid format.
+     *
+     * @param loanRequest loan request from user.
+     * @throws LoanRequestException if one of required parameters is null or parameters have incorrect format.
+     * @see LoanRequestException.ExceptionCause
+     */
+    private void validateLoanApplicationRequestDTO(ModelsLoanApplicationRequestDTO loanRequest) throws LoanRequestException {
+
+        if (logger.isTraceEnabled()) logger.trace("Validating loan request...");
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java:61]: Здесь и в других местах условие IF не имеет смыла, так как log.trace в любом случае выведет данные только в том случае, есть уровень логгирования не выше TRACE

File: Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+        if (loanRequest.getAmount() == null ||
+                loanRequest.getTerm() == null ||
+                loanRequest.getFirstName() == null ||
+                loanRequest.getLastName() == null ||
+                loanRequest.getEmail() == null ||
+                loanRequest.getBirthdate() == null ||
+                loanRequest.getPassportNumber() == null ||
+                loanRequest.getPassportSeries() == null) {
+            if (logger.isDebugEnabled()) logger.debug("One of loan request required parameters is null.");
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java:98]: См. выше коммент про логгирование, в других местах тоже лучше исправить

File: Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+        if (logger.isTraceEnabled()) logger.trace("Loan request is valid.");
+    }
+
+    /**
+     * Method performs null-check of {@link ModelsLoanApplicationRequestDTO} required parameters.
+     *
+     * @param loanRequest incoming loan request.
+     * @throws LoanRequestException if one of required parameters is null.
+     */
+    private void parametersNullCheck(ModelsLoanApplicationRequestDTO loanRequest) throws LoanRequestException {
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java:88]: Для таких целей есть ломбоковская аннотация @NotNull

File: Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java (hunk 8 above)
Comment (alxdv97) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java:98]: Еще один плюс использования ломбоковской аннотации @NotNull - там будет понятно, какое именно поле пришло null. В данном сообщении не понятно, что же именно было null

Comment (alxdv97) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java:98]: а вообще, ты же модели через open-api спецификацию задаешь? Там можно указать нуллабельность/обязательность и в том числе паттерны, например, для имени, чтобы не писать  самому методы вроде validateName()

File: Conveyor/src/main/java/ru/leonov/conveyor/service/ScoringService.java [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+     *
+     * @param scoringData data to perform scoring.
+     * @return credit rate.
+     * @throws ScoringException if scoring data is unacceptable to get credit.
+     */
+    private double calculateRate(ModelsScoringDataDTO scoringData) throws ScoringException {
+
+        if (logger.isTraceEnabled()) logger.trace("Calculating credit rate. Base rate is {}.", baseRate);
+
+        double resultRate = baseRate;
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/ScoringService.java:61]: В таких вычислениях лучше всегда использовать BigDecimal. Вообще, когда речь идет про деньги, тип Double нигде лучше не использовать

File: OpenApi/CreditConveyorApi.yaml [hunk 11]
```diff
@@ -0,0 +1,457 @@
...
+          description: Имя.
+        lastName:
+          type: string
+          description: Фамилия.
+        middleName:
+          type: string
+          description: Отчество.
+        email:
+          type: string
+          description: Электронная почта.
```
Comment (alxdv97) [inline OpenApi/CreditConveyorApi.yaml:239]: Для каждого такого поля всегда лучше прописывать некую валидацию, например pattern (для чисел min/max и тд)
Еще для всех полей лучше указать example, чтобы потом в сваггер это автоматически устанавливалось и можно было бы быстрее выполнять клиентский путь

Comment (alxdv97) [reply OpenApi/CreditConveyorApi.yaml:239]: А, ты указывал при объявлении запроса. Ну, так тоже можно

Comment (alxdv97) [reply OpenApi/CreditConveyorApi.yaml:239]: Просто удобнее читать информацию о том, каким может быть поле в том месте, где оно и объявляется, поэтому в следующий раз лучше писать примеры при объявлении полей, а не запросов

File: Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+
+    /**
+     * Validation of passport number.
+     *
+     * @param number customers passport number.
+     * @throws LoanRequestException if passport number not match required pattern.
+     */
+    private void validatePassportNumber(String number) throws LoanRequestException {
+        if (!number.matches("[\\d]{6}")) {
+            LoanRequestException exception = new LoanRequestException(
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java:208]: Обычно не создают объект экспешена перед тем как его кинуть)))
В данном случае можно сначала залоггировать с человеко-читаемым сообщением проблему (текст можно взять оттуда же, из энума эксепшена), а потом просто сразу кинуть эксепшен

File: Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+
+            resultList.add(modelsLoanOfferDTO);
+        }
+        resultList.get(1).setIsInsuranceEnabled(true);
+        resultList.get(2).setIsSalaryClient(true);
+        resultList.get(3).setIsSalaryClient(true);
+        resultList.get(3).setIsInsuranceEnabled(true);
+
+        //calculating different offers
+        for (ModelsLoanOfferDTO modelsLoanOfferDTO : resultList) {
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java:56]: Достаточно назвать сущность просто как LoanOfferDTO, потому что слово Models в данном контексте смысловой нагрузки не несет

File: Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+
+            double paymentForInsurance = 0;
+
+            //calculating current rate depending on base rate and booleans
+            if (Boolean.TRUE.equals(modelsLoanOfferDTO.getIsInsuranceEnabled())) {
```
[the remark this answers is not in the document]
Comment (cblpope3) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java:61]: В modelsLoanOfferDTO классе это булево поле реализовано не примитивным типом boolean, а ссылочным Boolean, и может быть null. Соответственно, это такая защита от null. Нужно её убрать?

File: Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java (hunk 3 above)
Comment (cblpope3) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java:143]: При делении и возведении в степень нужно указывать, иначе может кинуть исключение, либо считать неправильно. При остальных операциях, насколько я понял, не обязательно. Убрать везде, где уже написал?

File: Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java (hunk 4 above)
Comment (cblpope3) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java:248]: Посмотрел по коду, setScale используется 4 раза в конце методов "сосчитать ежемесячный платёж", "сосчитать платёж по процентам", "сосчитать ПСК" и "сосчитать ПСК по упрощённой формуле" (этот метод пока оставил, но пометил @Deprecated). Думаю, что это можно оставить как есть. Нужно ли в остальных случаях явно указывать скейл 8-10?

File: Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java (hunk 7 above)
Comment (cblpope3) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java:61]: А если внутри сообщения есть параметры, например `log.trace("Calculating monthly payment for {} roubles {}% credit for {} months.", creditAmount, yearlyCreditRate, creditTerm);`, лучше делать без if, или с ним? Линтеры как раз ругаются на подобные вызовы логгеров без if.

File: Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java (hunk 12 above)
Comment (cblpope3) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java:208]: Я сделал так, потому что при создании исключения я передаю в его конструктор параметр, при котором исключение произошло (например, при каком конкретно возрасте пользователь не прошёл проверку). Но, в принципе, это не так и нужно, поэтому заменил на сообщение без параметра.

File: Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java (hunk 13 above)
Comment (cblpope3) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java:56]: Похоже, что не получится убрать слово Models из названия dto файлов. Так как будет несколько микросервисов с отдельным api для каждого, но с общими dto, я вынес модели в отдельный файл Models.yaml, на который ссылается каждый ___Api.yaml файл. Слово "Models" openapi-генератор добавляет в название файла сам, так как оно прописано в ConveyorApi.yaml (например, ссылка на модель LoanOffer выглядит так: `$ref: '#/components/schemas/Models_LoanOffer')`, где 'Models_' - ссылка на файл Models.yaml).

File: Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java (hunk 8 above)
Comment (cblpope3) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java:98]: Полдня сегодня провозился с автоматической генерацией валидации параметров модели из openapi спецификации (проверка минимального/максимального значения, регулярные выражения и т.д.). В общем, условия валидации в спецификации прописаны, но openapi-generator генерирует только @NonNull аннотацию, других проверок не делает. Может быть есть какая-то настройка, которая решит проблему, я не нашёл...

File: Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java (hunk 13 above)
Comment (alxdv97) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java:70]: Да, вообще, по моему опыту проще объявление API и DTO класть в один файл, по трем причинам: 1. Фактические API - контракт взаимодействия, частями которого являются REST-эндпоинты и DTO. 2. удобная навигация по yaml одному yaml файлу 3. Не нужно  ссылаться на отдельный файл - не появляется ненужная приставка перед словом

File: Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (alxdv97) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/CreditCalculationService.java:75]: Можно и убрать, ведь если там null, то условие точно не выполнится. Если ты в данном случае переживаешь про NPE - ты же не вызываешь никаких методов на у того, что получилось в результате  выполнения метода getIsInsuranceEnabled(), соответственно NPE никогда не выкинется.

File: Conveyor/src/main/java/ru/leonov/conveyor/controller/Conveyor.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+@RestController
+@SuppressWarnings("unused")
+public class Conveyor implements DefaultApi {
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    private final PreScoringService preScoringService;
+    private final ScoringService scoringService;
+
+    @Autowired
+    public Conveyor(PreScoringService preScoringService, ScoringService scoringService) {
```
[the remark this answers is not in the document]
Comment (alxdv97) [reply Conveyor/src/main/java/ru/leonov/conveyor/controller/Conveyor.java]: Да, это компромиссный момент, если хочешь читаемость - делаешь без final и с RequiredArgsConstructor, хочешь безопасность - делаешь с final, на твоё усмотрение)

File: Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java (hunk 8 above)
Comment (alxdv97) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java]: Попробуй выставить настройку в помнике <useBeanValidation>false</useBeanValidation> в true - должно появиться аннотации вроде этих, так же на интерфейсах контроллеров появится аннотация validated
image

Comment (alxdv97) [reply Conveyor/src/main/java/ru/leonov/conveyor/service/PreScoringService.java]: Вообще, я недавно давно осознал одну мудрость - если взять все подряд подобные настройки и конфигурации из каких-нибудь туториалов, очень велик шанс того, что одна из этих настроек может сильно подпортить жизнь))) Особенно, когда докеры всякие конфигурируешь: просто перекатал не глядя все настройки из туториала, а потом нихрена не заводится)))) как говориться "плавали - знаем"))
Короче, мысль вот в чем: если используешь какие-то настройки, взятые из сети, лучше пройтись по каждой из них и понять, что именно она делает.
В твоем случае, вот ссылка, там описаны все проперти https://openapi-generator.tech/docs/generators/java/

File: Conveyor/src/main/java/ru/leonov/conveyor/controller/ConveyorController.java [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    private final PreScoringService preScoringService;
+    private final ScoringService scoringService;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public ResponseEntity<ModelsCreditDTO> postConveyorCalculation(ModelsScoringDataDTO modelsScoringDataDTO) {
+
+        if (log.isTraceEnabled()) log.trace("Got /conveyor/calculation request.");
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/controller/ConveyorController.java:35]: В прошлый раз обсуждали это - не имеет смысла перед каждым логом проверять уровень логгирования, он и так проверяется самим логгером. 
Иначе представь, какой бы это был неудобный инструмент, если бы так каждый раз нужно было бы делать)))
Пройдись по всем местам, посмотри, где еще проверяешь уровень логгирования перед написанием лога

File: Conveyor/src/main/java/ru/leonov/conveyor/controller/Conveyor.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (cblpope3) [reply Conveyor/src/main/java/ru/leonov/conveyor/controller/Conveyor.java]: Оставлю с final. Так и тестировать удобней, spring-context не нужен.

File: Conveyor/src/main/java/ru/leonov/conveyor/CreditConveyorApplication.java [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class CreditConveyorApplication {
+
+    public static void main(String[] args) {
+        SpringApplication.run(CreditConveyorApplication.class, args);
+        //todo remove this system.out
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/CreditConveyorApplication.java:11]: Думаю, это можно удалить)

File: Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,509 @@
...
+    private static final BigDecimal SALARY_TO_LOAN_RATE_MAX = BigDecimal.valueOf(20);
+
+
+    private static final double BASE_RATE = 15;
+    private ScoringService scoringService;
+    private ScoringDataDTO baseRateScoringRequest = new ScoringDataDTO();
+
+    @BeforeEach
+    void SetUp() {
+        scoringService = new ScoringService(BASE_RATE);
```
Comment (alxdv97) [inline Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java:78]: Почему не через spring, а через обычную джаву? В данном контексте конечно особо не важно, у тебя только ставку надо внедрить, но в дальнейшем лучше через спринг тестить, потому что там может быть миллион зависимостей у бина, а у этих зависимостей - свои зависимости, всех не создашь

Comment (alxdv97) [reply Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java:78]: И да, почитай про Mockito, чтобы уметь мокать сервисы, и реализуй Mockito в своих тестах. Там можно задавать объекту определенное поведение в зависимости от тесткейса

Comment (alxdv97) [reply Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java:78]: С помощью мокито можно указывать конкретное поведение мок-объекта и, например, проверять, был ли вызван метод у этого объекта необходимое число раз

File: Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,509 @@
...
+    // Женщина, возраст от 35 до 60 лет → ставка уменьшается на 3
+    @Test
+    void calculateCreditRateFemaleFineAge() {
+        //testing four cases to inspect all possible boundary conditions
+
+        baseRateScoringRequest.setGender(ScoringDataDTO.GenderEnum.FEMALE);
+        BigDecimal expectedRate = BigDecimal.valueOf(BASE_RATE + FEMALE_PREFERRED_AGE_CORRECTION);
+        BigDecimal calculatedRate;
+
+        try {
```
Comment (alxdv97) [inline Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java:168]: в тестах try-catch не нужен, можно указать ожидаемое исключение в аннотации перед тесткейсом

File: Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,509 @@
...
+
+        EmploymentDTO employmentDTO = baseRateScoringRequest.getEmployment();
+        employmentDTO.setWorkExperienceTotal(MIN_TOTAL_EXPERIENCE - 1);
+        baseRateScoringRequest.setEmployment(employmentDTO);
+
+        try {
+            scoringService.calculateRate(baseRateScoringRequest);
+        } catch (ScoringException e) {
+            assertEquals(expectedExceptionMessage, e.getMessage());
+            ++numberOfHappenedExceptions;
```
Comment (alxdv97) [inline Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java:379]: Мысль с numberOfHappenedExceptions понятна, но получается, что ты делаешь какую-то мудрёную логику там, где ее не должно быть (в тестах). Тесты всегда должны быть максмально простые: конфигурация моков в before, а дальше просто пишешь много "плоских" тестов, которые только проверяют ожидаемое и действительное значение  на основании тестовых данных. Выкидывание экспешена в ходе выполнение теста - такой же обычный результат выполнения, как и нормальное завершение метода, на выкидывание эксепшена тоже нужно писать тесты, указав эксепшен в аннотации test в параметре expected ЕМНИП

File: Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java (hunk 18 above)
Comment (cblpope3) [reply Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java]: > Почему не через spring, а через обычную джаву? В данном контексте конечно особо не важно, у тебя только ставку надо внедрить, но в дальнейшем лучше через спринг тестить, потому что там может быть миллион зависимостей у бина, а у этих зависимостей - свои зависимости, всех не создашь))

Сделал через спринг. Причём решил, что все параметры скоринга должны быть заданы в application.yaml файле и инжектиться при помощи @Value аннотации, так как получается избыточное дублирование данных и в контроллер и в тест для него. Мысль правильная?

Comment (cblpope3) [reply Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java]: > И да, почитай про Mockito, чтобы уметь мокать сервисы, и реализуй Mockito в своих тестах. Там можно задавать объекту определенное поведение в зависимости от тесткейса

Я в данном случае не понял, куда тут прилепить Мокито? Этот сервис конечный и не имеет других сервисов внутри себя, так что и мокать нечего. Я использовал Мокито, например, в тестах контроллеров. Там нужно было мокать сервисы, которые использует контроллер.

File: Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java (hunk 19 above)
Comment (cblpope3) [reply Conveyor/src/test/java/ru/leonov/conveyor/service/ScoringServiceTest.java]: try-catch заменил на assertThrows, так сойдёт? Мне показалось это более правильным и гибким.

File: Conveyor/src/main/java/ru/leonov/conveyor/service/ScoringService.java [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+ * This service handle scoring process.
+ */
+@Slf4j
+@Service
+public class ScoringService {
+
+    private final BigDecimal salaryToLoanRateLimit;
+    private final BigDecimal baseRate;
+
+    @SuppressWarnings("unused")
```
Comment (alxdv97) [inline Conveyor/src/main/java/ru/leonov/conveyor/service/ScoringService.java:25]: Не помню точно, но вроде можно было эту аннотацию на весь класс повесить
